packages/@aws-cdk/assertions/lib/assertions.ts

```diff
@@ -53,23 +54,49 @@ export class TemplateAssertions {
   /**
    * Assert that a resource of the given type and properties exists in the
    * CloudFormation template.
+   * By default, performs partial matching on the `Properties` key of the resource, via the
+   * `Match.objectLike()`. To configure different behavour, use other matchers in the `Match` class.
    * @param type the resource type; ex: `AWS::S3::Bucket`
    * @param props the 'Properties' section of the resource as should be expected in the template.
    */
   public hasResourceProperties(type: string, props: any): void {
-    const assertion = assert.haveResource(type, props, assert.ResourcePart.Properties);
-    assertion.assertOrThrow(this.inspector);
+    this.hasResource(type, Match.objectLike({
+      Properties: Match.isMatcher(props) ? props : Match.objectLike(props, { partial: true }),
+    }, { partial: true }));
   }
 
   /**
    * Assert that a resource of the given type and given definition exists in the
    * CloudFormation template.
+   * By default, performs partial matching on the resource, via the `Match.objectLike()`.
+   * To configure different behavour, use other matchers in the `Match` class.
    * @param type the resource type; ex: `AWS::S3::Bucket`
    * @param props the entire defintion of the resource as should be expected in the template.
    */
-  public hasResourceDefinition(type: string, props: any): void {
-    const assertion = assert.haveResource(type, props, assert.ResourcePart.CompleteDefinition);
-    assertion.assertOrThrow(this.inspector);
+  public hasResource(type: string, props: any): void {
+    const matcher = Match.isMatcher(props) ? props : Match.objectLike(props, { partial: true });
+    let closestResult: MatchFailure[] | undefined = undefined;
+    let count: number = 0;
+    for (const logicalId of Object.keys(this.inspector.value.Resources || {})) {
+      const resource = this.inspector.value.Resources[logicalId];
+      if (resource.Type === type) {
+        count++;
+        const result = matcher.test(resource);
+        if (result.length === 0) {
+          return;
+        }
+        if (closestResult === undefined || closestResult.length > result.length) {
```

Comment:
This whole notion of a "closest match" (and so a "size"/"distance" concept on a match result) needs to be part of the matchers library and needs to be reused in a couple of places.

Reply:
I suppose you're suggesting turning `MatchFailure` into a class and hiding all of these behind it, yes?

Reply:
Yes. Although I would call it `MatchResult`.